src/bootstrap/lib.rs

```diff
@@ -692,7 +692,7 @@ impl Build {
             }
         } else {
             let base = self.llvm_out(self.config.build).join("build");
-            let base = if !self.ninja() && self.config.build.contains("msvc") {
+            let base = if !self.maybe_ninja() && self.config.build.contains("msvc") {
```

Comment:
Hm.. This feels a bit off. Who is calling llvm_filecheck to cause problems here? Should we be giving them a Option<PathBuf>? Will they run into trouble?

Reply:
Here's the backtrace:
```
   1: bootstrap::Build::ninja
             at ./src/bootstrap/lib.rs:1366:17
   2: bootstrap::Build::llvm_filecheck
             at ./src/bootstrap/lib.rs:695:28
   3: bootstrap::sanity::check
             at ./src/bootstrap/sanity.rs:161:25
   4: bootstrap::Build::new
             at ./src/bootstrap/lib.rs:448:9
   5: bootstrap::main
             at ./src/bootstrap/bin/main.rs:33:5
```
which is used like so:
```rust
        // Externally configured LLVM requires FileCheck to exist
        let filecheck = build.llvm_filecheck(build.build);
        if !filecheck.starts_with(&build.out) && !filecheck.exists() && build.config.codegen_tests {
            panic!("FileCheck executable {:?} does not exist", filecheck);
        }
```
So the only way this can go wrong is if using system LLVM. But this branch is never taken for system LLVM, it will be caught above on line 673:
```rust
        } else if let Some(s) = target_config.and_then(|c| c.llvm_config.as_ref()) {
```
-

Reply:
Also, in general if `llvm_filecheck` is wrong, it will be very clear because the file just won't be found and the command will fail.

Reply:
must_have cmake also doesn't work, it requires cmake even when LLVM isn't being built:
```
thread 'main' panicked at '

couldn't find required command: "cmake"

', src\bootstrap\sanity.rs:60:13
```
I think a fix might be to move all these checks to `native.rs` instead? I don't quite understand the point of sanity.rs.

Reply:
We want the build to fail early if it's going to fail, to avoid running for a long time and then failing (when someone has stepped away, for example, from their computer).

I don't quite understand what you mean by must_have cmake not working there. Note that it's required for both LLVM but also if you're enabling any of the sanitizers; maybe you have those enabled?

Reply:
> I don't quite understand what you mean by must_have cmake not working there. Note that it's required for both LLVM but also if you're enabling any of the sanitizers; maybe you have those enabled?

No, I do not have any sanitizers enabled. The problem is the very first time someone runs `x.py setup` they will have no config.toml, so x.py thinks it needs to build LLVM from source. The fixes I can think of are:
1. Make downloading LLVM the default, so that `x.py setup` is confused in the other direction and doesn't require cmake no matter what. This seems ok, since `setup` doesn't build any code (although of course changing the default would probably need an MCP or something).
2. Never run `sanity` when running `x.py setup`, since it's not building any code.
3. Move the check from `sanity` to `native`. You said above this defeats the point since it will fail slowly instead of quickly.
4. Somehow reload the config after `x.py setup` generates it.

My preferred fix is 2, I'll work on implementing that.